File: src/main/java/io/scriptor/fdt/Buildable.java
package io.scriptor.fdt;

import org.jetbrains.annotations.NotNull;

import java.util.function.Consumer;
import java.util.function.Function;

public interface Buildable<T> {

    T build();

    default void build(final @NotNull Consumer<T> consumer) {
        consumer.accept(build());
    }

    default <V> V map(final @NotNull Function<T, V> mapper) {
        return mapper.apply(build());
    }
}

File: src/main/java/io/scriptor/fdt/Buildable.kt
package io.scriptor.fdt

import java.util.function.Consumer
import java.util.function.Function

interface Buildable<T> {

    fun build(): T

    fun build(consumer: Consumer<T>) {
        consumer.accept(build())
    }

    fun <V> map(mapper: Function<T, V>): V {
        return mapper.apply(build())
    }
}
